kong/pdk/vault.lua

```diff
@@ -54,32 +59,143 @@ local BRACE_END = byte("}")
+--   config   = {       -- if there are any config options specified
+--     prefix = "SSL_"
+--   },
+--   version  = 1       -- if the version is specified
+-- }
+-- ```
+--
+-- @local
+-- @function parse_reference
+-- @tparam string reference reference to parse
+-- @treturn table|nil a table containing each component of the reference, or `nil` on error
+-- @treturn string|nil error message on failure, otherwise `nil`
+local function parse_reference(reference)
+  if not is_reference(reference) then
+    return nil, fmt("not a reference [%s]", tostring(reference))
+  end
+
+  local url, err = parse_url(sub(reference, 2, -2))
+  if not url then
+    return nil, fmt("reference is not url (%s) [%s]", err, reference)
+  end
+
+  local name = url.host
+  if not name then
+    return nil, fmt("reference url is missing host [%s]", reference)
+  end
+
+  local path = url.path
+  if not path then
+    return nil, fmt("reference url is missing path [%s]", reference)
+  end
+
+  local resource = sub(path, 2)
+  if resource == "" then
+    return nil, fmt("reference url has empty path [%s]", reference)
+  end
+
+  local version = url.fragment
+  if version then
+    version = tonumber(version, 10)
+    if not version then
+      return nil, fmt("reference url has invalid version [%s]", reference)
+    end
+  end
+
+  local key
+  local parts = parse_path(resource)
+  local count = #parts
+  if count == 1 then
+    resource = unescape_uri(parts[1])
+
+  else
+    resource = unescape_uri(concat(parts, "/", 1, count - 1))
+    if parts[count] ~= "" then
+      key = unescape_uri(parts[count])
+    end
+  end
+
+  if resource == "" then
+    return nil, fmt("reference url has invalid path [%s]", reference)
+  end
+
+  local config
+  local query = url.query
+  if query and query ~= "" then
+    config = decode_args(query)
+  end
+
+  return {
+    name = url.host,
+    resource = resource,
+    key = key,
+    config = config,
+    version = version,
+  }
+end
+
+
+---
+-- Create a instance of PDK Vault module
+--
+-- @local
+-- @function new
+-- @tparam table self a PDK instance
+-- @treturn table a new instance of Vault
 local function new(self)
   -- Don't put this onto the top level of the file unless you're prepared for a surprise
   local Schema = require "kong.db.schema"
 
-  local ROTATION_SEMAPHORE = semaphore.new(1)
-  local ROTATION_WAIT = 0
+  local ROTATION_MUTEX_OPTS = {
+    name = "vault-rotation",
+    exptime = ROTATION_INTERVAL * 1.5, -- just in case the lock is not properly released
+    timeout = 0, -- we don't want to wait for release as we run a recurring timer
+  }
 
   local LRU = lrucache.new(1000)
-  local SHDICT = ngx.shared.kong_secrets
+  local RETRY_LRU = lrucache.new(1000)
 
-  local KEY_BUFFER = buffer.new(100)
+  local SECRETS_CACHE = ngx.shared.kong_secrets
 
-  local RETRY_LRU = lrucache.new(1000)
-  local RETRY_SEMAPHORE = semaphore.new(1)
-  local RETRY_WAIT = 1
-  local RETRY_TTL = 10
+  local KEY_BUFFER = buffer.new(100)
```

Comment:
Is the benefit of using a shared `KEY_BUFFER` instead of allocating one as required actually measureable?  Is it an arbitrary optimization attempt or is there evidence that it has an impact?

If it is certain that the optimization makes sense, I'd still suggest to use two separate statically allocated buffers for the two places in which `KEY_BUFFER` Is currently used, and put the declaration into a `do` block scoped to the function that requires the buffer.

Reply:
I think it won't matter that much. The rule of thumb: If you know it is rather small buffer, then reusing buffer means that we don't have to memory allocate. We just use it. With large buffers this creates persistent memory usage which can be bad so it is advisable not use it with large concatenation operations. Whether it makes sense to optimize small strings is a good question, probably not, but that's the reason I think for this code. I will take another look at it.

Reply:
I made changes to this too. In other place I used `buffer.new():set()` and in other function I moved reusable buffer closer to code.